src/vs/workbench/contrib/notebook/browser/viewModel/notebookViewModel.ts

```diff
@@ -228,14 +230,27 @@ export class NotebookViewModel extends Disposable implements EditorFoldingStateD
 		private _layoutInfo: NotebookLayoutInfo | null,
 		@IInstantiationService private readonly _instantiationService: IInstantiationService,
 		@IBulkEditService private readonly _bulkEditService: IBulkEditService,
-		@IUndoRedoService private readonly _undoService: IUndoRedoService
+		@IUndoRedoService private readonly _undoService: IUndoRedoService,
+		@IModelService modelService: IModelService,
 	) {
 		super();
 
 		MODEL_ID++;
 		this.id = '$notebookViewModel' + MODEL_ID;
 		this._instanceId = strings.singleLetterHash(MODEL_ID);
 
+		this._register(modelService.onModelAdded(e => {
```

Comment:
Unsure if the view model is the right place or if the listeners should be in the notebook text model. For instance, an extension can open a notebook document without an editor and the change the language for a cell. I believe in that case we would miss those updates but I might be wrong

Reply:
is there any guarantee of what the cell URI will be before the notebook is opened? I will check what it does in that case

Reply:
The URI is always the same and there is a guarantee that the [notebook is opened](https://github.com/microsoft/vscode/blob/216bff5771527f67800b58cee745ed8217c07c9b/src/vs/workbench/contrib/notebook/browser/notebook.contribution.ts#L524) before (or at the same time) the cell document is opened. Tho, there is no guarantee that an editor is around and therefore likely no view model